Separate android artifact

JeroenMols opened the pull request:
Added a separate Android artifact to backport `Base64` to Android versions older than API 26.
Fixes https://github.com/streem/pbandk/issues/75

-------------------------------

What this PR does is:

- Create a new shared `SourceSet` for `Jvm` and `Android` sources
- Create a separate `Util` implementation for `Jvm` (using `java.util.Base64`) and `Android` (using `android.util.Base64`)
- Add a new Maven Artifact `pbandk-runtime-android`
- Minimum Android version 14 -> same as Jetpack/support library
- Target/Compile SDK 30 -> latest stable
- **Update**: move all Java `jvmTest` source files to separate module
- **Update**: use the `Jvm` `Util` implementation in Android tests

Note that all unit tests are still run on the `JVM` target, because we need a `Base64` implementation to run those tests (`expected` needs to be provided with an `actual`) and the `android.util.Base64` will throw a `NotMockedException` in unit tests.

Tested this by running all conformance tests and creating all Maven artifacts in my local maven repository.

```text
└── pro
        └── streem
            └── pbandk
                ├── pbandk-runtime
                │   ├── 0.10.0-SNAPSHOT
                │   │   ├── maven-metadata-local.xml
                │   │   ├── pbandk-runtime-0.10.0-SNAPSHOT.jar
                │   │   ├── pbandk-runtime-0.10.0-SNAPSHOT.module
                │   │   └── pbandk-runtime-0.10.0-SNAPSHOT.pom
                │   └── maven-metadata-local.xml
                ├── pbandk-runtime-android
                │   ├── 0.10.0-SNAPSHOT
                │   │   ├── maven-metadata-local.xml
                │   │   ├── pbandk-runtime-android-0.10.0-SNAPSHOT-sources.jar
                │   │   ├── pbandk-runtime-android-0.10.0-SNAPSHOT.aar
                │   │   ├── pbandk-runtime-android-0.10.0-SNAPSHOT.module
                │   │   └── pbandk-runtime-android-0.10.0-SNAPSHOT.pom
                │   └── maven-metadata-local.xml
                ├── pbandk-runtime-iosarm64
                │   ├── 0.10.0-SNAPSHOT
                │   │   ├── maven-metadata-local.xml
                │   │   ├── pbandk-runtime-iosarm64-0.10.0-SNAPSHOT-sources.jar
                │   │   ├── pbandk-runtime-iosarm64-0.10.0-SNAPSHOT.klib
                │   │   ├── pbandk-runtime-iosarm64-0.10.0-SNAPSHOT.module
                │   │   └── pbandk-runtime-iosarm64-0.10.0-SNAPSHOT.pom
                │   └── maven-metadata-local.xml
                ├── pbandk-runtime-iosx64
                │   ├── 0.10.0-SNAPSHOT
                │   │   ├── maven-metadata-local.xml
                │   │   ├── pbandk-runtime-iosx64-0.10.0-SNAPSHOT-sources.jar
                │   │   ├── pbandk-runtime-iosx64-0.10.0-SNAPSHOT.klib
                │   │   ├── pbandk-runtime-iosx64-0.10.0-SNAPSHOT.module
                │   │   └── pbandk-runtime-iosx64-0.10.0-SNAPSHOT.pom
                │   └── maven-metadata-local.xml
                ├── pbandk-runtime-js
                │   ├── 0.10.0-SNAPSHOT
                │   │   ├── maven-metadata-local.xml
                │   │   ├── pbandk-runtime-js-0.10.0-SNAPSHOT-sources.jar
                │   │   ├── pbandk-runtime-js-0.10.0-SNAPSHOT.jar
                │   │   ├── pbandk-runtime-js-0.10.0-SNAPSHOT.module
                │   │   └── pbandk-runtime-js-0.10.0-SNAPSHOT.pom
                │   └── maven-metadata-local.xml
                ├── pbandk-runtime-jvm
                │   ├── 0.10.0-SNAPSHOT
                │   │   ├── maven-metadata-local.xml
                │   │   ├── pbandk-runtime-jvm-0.10.0-SNAPSHOT-sources.jar
                │   │   ├── pbandk-runtime-jvm-0.10.0-SNAPSHOT.jar
                │   │   ├── pbandk-runtime-jvm-0.10.0-SNAPSHOT.module
                │   │   └── pbandk-runtime-jvm-0.10.0-SNAPSHOT.pom
                │   └── maven-metadata-local.xml
                ├── pbandk-runtime-linuxx64
                │   ├── 0.10.0-SNAPSHOT
                │   │   ├── maven-metadata-local.xml
                │   │   ├── pbandk-runtime-linuxx64-0.10.0-SNAPSHOT-sources.jar
                │   │   ├── pbandk-runtime-linuxx64-0.10.0-SNAPSHOT.klib
                │   │   ├── pbandk-runtime-linuxx64-0.10.0-SNAPSHOT.module
                │   │   └── pbandk-runtime-linuxx64-0.10.0-SNAPSHOT.pom
                │   └── maven-metadata-local.xml
                ├── pbandk-runtime-macosx64
                │   ├── 0.10.0-SNAPSHOT
                │   │   ├── maven-metadata-local.xml
                │   │   ├── pbandk-runtime-macosx64-0.10.0-SNAPSHOT-sources.jar
                │   │   ├── pbandk-runtime-macosx64-0.10.0-SNAPSHOT.klib
                │   │   ├── pbandk-runtime-macosx64-0.10.0-SNAPSHOT.module
                │   │   └── pbandk-runtime-macosx64-0.10.0-SNAPSHOT.pom
                │   └── maven-metadata-local.xml
                └── pbandk-runtime-metadata
                    ├── 0.10.0-SNAPSHOT
                    │   ├── maven-metadata-local.xml
                    │   ├── pbandk-runtime-metadata-0.10.0-SNAPSHOT-sources.jar
                    │   ├── pbandk-runtime-metadata-0.10.0-SNAPSHOT.jar
                    │   ├── pbandk-runtime-metadata-0.10.0-SNAPSHOT.module
                    │   └── pbandk-runtime-metadata-0.10.0-SNAPSHOT.pom
                    └── maven-metadata-local.xml
```

JeroenMols commented on buildSrc/src/main/kotlin/Versions.kt:
```diff
@@ -8,4 +8,7 @@ object Versions {
     const val protobufJs = "6.10.2"
     const val springBootGradlePlugin = "2.3.7.RELEASE"
     const val osDetectorGradlePlugin = "1.6.2"
+    const val androidGradlePlugin = "4.1.2"
```
There is an argument to be made that `androidMinSdk` is actually a configuration rather than a version, but decided to keep it here for simplicity. Happy to move this around.

JeroenMols commented on runtime/build.gradle.kts:
```diff
@@ -2,15 +2,25 @@ import org.jetbrains.kotlin.gradle.plugin.mpp.KotlinNativeTarget
-    jvm {
-        withJava()
+    android {
+        publishLibraryVariants("release")
```
Weirdly, the publishing of `android` artifacts seems to work differently than the other artifacts. 😢 
This also causes duplicate code to generate the artifact name.

garyp replied:
Oh, I think this is because the `publishing {}` block at the bottom of this file needs to be wrapped inside of `afterEvaluate {}` to work correctly for Android. See https://developer.android.com/studio/build/maven-publish-plugin for more details.

JeroenMols replied:
I've tried your suggestion, but that didn't work. Which is expected as per the documentation here:
https://kotlinlang.org/docs/mpp-publish-lib.html#publish-an-android-library

I think this is because Android has variants by default and KMP wouldn't know what variants to publish unless some configuration tells it.

garyp replied:
I misunderstood which line of the code your original comment was referring to. The `publishLibraryVariants("release")` is indeed required. I meant that you can remove this block:

```kt
        mavenPublication {
            this.artifactId = "pbandk-$artifactId"
        }
```

if you wrap the `publishing {}` block in `afterEvaluate`.

JeroenMols replied:
Yeah I tried that as well. IIRC: the Android artifact doesn't have an android suffix in its name.

So we could follow your approach, but then we would have to rename the Android artifact based on some metadata in the `publishing` block.

garyp replied:
Are you sure? I tried `./gradlew publishToMavenLocal` with your PR, and then I tried the same command with my changes applied on top of your PR, and both commands generated the same set of files (I made sure to delete my local maven repo in-between).

Here's the change I made:

```diff
diff --git a/runtime/build.gradle.kts b/runtime/build.gradle.kts
index 1aacbec..8a8788f 100644
--- a/runtime/build.gradle.kts
+++ b/runtime/build.gradle.kts
@@ -14,9 +14,6 @@ kotlin {
 
     android {
         publishLibraryVariants("release")
-        mavenPublication {
-            this.artifactId = "pbandk-$artifactId"
-        }
     }
 
     jvm()
@@ -170,13 +167,15 @@ tasks {
     }
 }
 
-publishing {
-    publications.withType<MavenPublication>().configureEach {
-        artifactId = "pbandk-$artifactId"
-        description = "Library for pbandk runtime protobuf code"
-        pom {
-            configureForPbandk()
+afterEvaluate {
+    publishing {
+        publications.withType<MavenPublication>().configureEach {
+            artifactId = "pbandk-$artifactId"
+            description = "Library for pbandk runtime protobuf code"
+            pom {
+                configureForPbandk()
+            }
+            addBintrayRepository(project, this)
         }
-        addBintrayRepository(project, this)
     }
 }
```

and here's the resulting set of published files:

```
pbandk-runtime
pbandk-runtime/0.10.0-SNAPSHOT
pbandk-runtime/0.10.0-SNAPSHOT/maven-metadata-local.xml
pbandk-runtime/0.10.0-SNAPSHOT/pbandk-runtime-0.10.0-SNAPSHOT.jar
pbandk-runtime/0.10.0-SNAPSHOT/pbandk-runtime-0.10.0-SNAPSHOT.module
pbandk-runtime/0.10.0-SNAPSHOT/pbandk-runtime-0.10.0-SNAPSHOT.pom
pbandk-runtime/maven-metadata-local.xml
pbandk-runtime-android
pbandk-runtime-android/0.10.0-SNAPSHOT
pbandk-runtime-android/0.10.0-SNAPSHOT/maven-metadata-local.xml
pbandk-runtime-android/0.10.0-SNAPSHOT/pbandk-runtime-android-0.10.0-SNAPSHOT-sources.jar
pbandk-runtime-android/0.10.0-SNAPSHOT/pbandk-runtime-android-0.10.0-SNAPSHOT.aar
pbandk-runtime-android/0.10.0-SNAPSHOT/pbandk-runtime-android-0.10.0-SNAPSHOT.module
pbandk-runtime-android/0.10.0-SNAPSHOT/pbandk-runtime-android-0.10.0-SNAPSHOT.pom
pbandk-runtime-android/maven-metadata-local.xml
pbandk-runtime-iosarm64
pbandk-runtime-iosarm64/0.10.0-SNAPSHOT
pbandk-runtime-iosarm64/0.10.0-SNAPSHOT/maven-metadata-local.xml
pbandk-runtime-iosarm64/0.10.0-SNAPSHOT/pbandk-runtime-iosarm64-0.10.0-SNAPSHOT-sources.jar
pbandk-runtime-iosarm64/0.10.0-SNAPSHOT/pbandk-runtime-iosarm64-0.10.0-SNAPSHOT.klib
pbandk-runtime-iosarm64/0.10.0-SNAPSHOT/pbandk-runtime-iosarm64-0.10.0-SNAPSHOT.module
pbandk-runtime-iosarm64/0.10.0-SNAPSHOT/pbandk-runtime-iosarm64-0.10.0-SNAPSHOT.pom
pbandk-runtime-iosarm64/maven-metadata-local.xml
pbandk-runtime-iosx64
pbandk-runtime-iosx64/0.10.0-SNAPSHOT
pbandk-runtime-iosx64/0.10.0-SNAPSHOT/maven-metadata-local.xml
pbandk-runtime-iosx64/0.10.0-SNAPSHOT/pbandk-runtime-iosx64-0.10.0-SNAPSHOT-sources.jar
pbandk-runtime-iosx64/0.10.0-SNAPSHOT/pbandk-runtime-iosx64-0.10.0-SNAPSHOT.klib
pbandk-runtime-iosx64/0.10.0-SNAPSHOT/pbandk-runtime-iosx64-0.10.0-SNAPSHOT.module
pbandk-runtime-iosx64/0.10.0-SNAPSHOT/pbandk-runtime-iosx64-0.10.0-SNAPSHOT.pom
pbandk-runtime-iosx64/maven-metadata-local.xml
pbandk-runtime-js
pbandk-runtime-js/0.10.0-SNAPSHOT
pbandk-runtime-js/0.10.0-SNAPSHOT/maven-metadata-local.xml
pbandk-runtime-js/0.10.0-SNAPSHOT/pbandk-runtime-js-0.10.0-SNAPSHOT-sources.jar
pbandk-runtime-js/0.10.0-SNAPSHOT/pbandk-runtime-js-0.10.0-SNAPSHOT.jar
pbandk-runtime-js/0.10.0-SNAPSHOT/pbandk-runtime-js-0.10.0-SNAPSHOT.module
pbandk-runtime-js/0.10.0-SNAPSHOT/pbandk-runtime-js-0.10.0-SNAPSHOT.pom
pbandk-runtime-js/maven-metadata-local.xml
pbandk-runtime-jvm
pbandk-runtime-jvm/0.10.0-SNAPSHOT
pbandk-runtime-jvm/0.10.0-SNAPSHOT/maven-metadata-local.xml
pbandk-runtime-jvm/0.10.0-SNAPSHOT/pbandk-runtime-jvm-0.10.0-SNAPSHOT-sources.jar
pbandk-runtime-jvm/0.10.0-SNAPSHOT/pbandk-runtime-jvm-0.10.0-SNAPSHOT.jar
pbandk-runtime-jvm/0.10.0-SNAPSHOT/pbandk-runtime-jvm-0.10.0-SNAPSHOT.module
pbandk-runtime-jvm/0.10.0-SNAPSHOT/pbandk-runtime-jvm-0.10.0-SNAPSHOT.pom
pbandk-runtime-jvm/maven-metadata-local.xml
pbandk-runtime-linuxx64
pbandk-runtime-linuxx64/0.10.0-SNAPSHOT
pbandk-runtime-linuxx64/0.10.0-SNAPSHOT/maven-metadata-local.xml
pbandk-runtime-linuxx64/0.10.0-SNAPSHOT/pbandk-runtime-linuxx64-0.10.0-SNAPSHOT-sources.jar
pbandk-runtime-linuxx64/0.10.0-SNAPSHOT/pbandk-runtime-linuxx64-0.10.0-SNAPSHOT.klib
pbandk-runtime-linuxx64/0.10.0-SNAPSHOT/pbandk-runtime-linuxx64-0.10.0-SNAPSHOT.module
pbandk-runtime-linuxx64/0.10.0-SNAPSHOT/pbandk-runtime-linuxx64-0.10.0-SNAPSHOT.pom
pbandk-runtime-linuxx64/maven-metadata-local.xml
pbandk-runtime-macosx64
pbandk-runtime-macosx64/0.10.0-SNAPSHOT
pbandk-runtime-macosx64/0.10.0-SNAPSHOT/maven-metadata-local.xml
pbandk-runtime-macosx64/0.10.0-SNAPSHOT/pbandk-runtime-macosx64-0.10.0-SNAPSHOT-sources.jar
pbandk-runtime-macosx64/0.10.0-SNAPSHOT/pbandk-runtime-macosx64-0.10.0-SNAPSHOT.klib
pbandk-runtime-macosx64/0.10.0-SNAPSHOT/pbandk-runtime-macosx64-0.10.0-SNAPSHOT.module
pbandk-runtime-macosx64/0.10.0-SNAPSHOT/pbandk-runtime-macosx64-0.10.0-SNAPSHOT.pom
pbandk-runtime-macosx64/maven-metadata-local.xml
pbandk-runtime-metadata
pbandk-runtime-metadata/0.10.0-SNAPSHOT
pbandk-runtime-metadata/0.10.0-SNAPSHOT/maven-metadata-local.xml
pbandk-runtime-metadata/0.10.0-SNAPSHOT/pbandk-runtime-metadata-0.10.0-SNAPSHOT-sources.jar
pbandk-runtime-metadata/0.10.0-SNAPSHOT/pbandk-runtime-metadata-0.10.0-SNAPSHOT.jar
pbandk-runtime-metadata/0.10.0-SNAPSHOT/pbandk-runtime-metadata-0.10.0-SNAPSHOT.module
pbandk-runtime-metadata/0.10.0-SNAPSHOT/pbandk-runtime-metadata-0.10.0-SNAPSHOT.pom
pbandk-runtime-metadata/maven-metadata-local.xml
protoc-gen-kotlin-jvm
protoc-gen-kotlin-jvm/0.10.0-SNAPSHOT
protoc-gen-kotlin-jvm/0.10.0-SNAPSHOT/maven-metadata-local.xml
protoc-gen-kotlin-jvm/0.10.0-SNAPSHOT/protoc-gen-kotlin-jvm-0.10.0-SNAPSHOT-jvm8.jar
protoc-gen-kotlin-jvm/0.10.0-SNAPSHOT/protoc-gen-kotlin-jvm-0.10.0-SNAPSHOT.pom
protoc-gen-kotlin-jvm/maven-metadata-local.xml
protoc-gen-kotlin-lib
protoc-gen-kotlin-lib/0.10.0-SNAPSHOT
protoc-gen-kotlin-lib/0.10.0-SNAPSHOT/maven-metadata-local.xml
protoc-gen-kotlin-lib/0.10.0-SNAPSHOT/protoc-gen-kotlin-lib-0.10.0-SNAPSHOT.jar
protoc-gen-kotlin-lib/0.10.0-SNAPSHOT/protoc-gen-kotlin-lib-0.10.0-SNAPSHOT.module
protoc-gen-kotlin-lib/0.10.0-SNAPSHOT/protoc-gen-kotlin-lib-0.10.0-SNAPSHOT.pom
protoc-gen-kotlin-lib/maven-metadata-local.xml
protoc-gen-kotlin-lib-jvm
protoc-gen-kotlin-lib-jvm/0.10.0-SNAPSHOT
protoc-gen-kotlin-lib-jvm/0.10.0-SNAPSHOT/maven-metadata-local.xml
protoc-gen-kotlin-lib-jvm/0.10.0-SNAPSHOT/protoc-gen-kotlin-lib-jvm-0.10.0-SNAPSHOT-sources.jar
protoc-gen-kotlin-lib-jvm/0.10.0-SNAPSHOT/protoc-gen-kotlin-lib-jvm-0.10.0-SNAPSHOT.jar
protoc-gen-kotlin-lib-jvm/0.10.0-SNAPSHOT/protoc-gen-kotlin-lib-jvm-0.10.0-SNAPSHOT.module
protoc-gen-kotlin-lib-jvm/0.10.0-SNAPSHOT/protoc-gen-kotlin-lib-jvm-0.10.0-SNAPSHOT.pom
protoc-gen-kotlin-lib-jvm/maven-metadata-local.xml
protoc-gen-kotlin-lib-metadata
protoc-gen-kotlin-lib-metadata/0.10.0-SNAPSHOT
protoc-gen-kotlin-lib-metadata/0.10.0-SNAPSHOT/maven-metadata-local.xml
protoc-gen-kotlin-lib-metadata/0.10.0-SNAPSHOT/protoc-gen-kotlin-lib-metadata-0.10.0-SNAPSHOT-sources.jar
protoc-gen-kotlin-lib-metadata/0.10.0-SNAPSHOT/protoc-gen-kotlin-lib-metadata-0.10.0-SNAPSHOT.jar
protoc-gen-kotlin-lib-metadata/0.10.0-SNAPSHOT/protoc-gen-kotlin-lib-metadata-0.10.0-SNAPSHOT.module
protoc-gen-kotlin-lib-metadata/0.10.0-SNAPSHOT/protoc-gen-kotlin-lib-metadata-0.10.0-SNAPSHOT.pom
protoc-gen-kotlin-lib-metadata/maven-metadata-local.xml
```

JeroenMols replied:
I had checked the artifact name by printing them in the `publishing` block, not the actual resulting files.

I'll try to check again over the weekend.

garyp replied:
Since this would be just a minor cleanup, I'll go ahead and merge the rest of this PR so we can kick off some testing of the new version internally. If you don't mind, please still check the publishing thing when you have a chance and put up another PR if my suggestion works out.

JeroenMols commented on runtime/build.gradle.kts:
```diff
@@ -2,15 +2,25 @@ import org.jetbrains.kotlin.gradle.plugin.mpp.KotlinNativeTarget
 kotlin {
 
-    jvm {
-        withJava()
```
I'm not sure why we needed the `withJava()` expression, but this caused a failure after adding the `android` sourceSet, because the Java plugin was being applied.

JeroenMols commented on runtime/build.gradle.kts:
```diff
@@ -50,7 +60,19 @@ kotlin {
             }
         }
 
+        val commonJvmAndroid = create("commonJvmAndroid") {
```
Common sourceSet that contains all code that used to be in the `jvm` sourceset, except the `Util` class

garyp replied:
According to https://kotlinlang.org/docs/mpp-share-on-platforms.html#configure-the-hierarchical-structure-manually:

> Kotlin doesn’t currently support sharing a source set for these combinations:
> * Several JVM targets
> * JVM + Android targets
> * Several JS targets

It seems that Gradle is happy building from the command line. But when I open a file from `commonJvmAndroid` in IntelliJ (e.g. `MessageExtensionsJvm.kt`), it fails to resolve the java-specific imports.

I'm not sure what to make of this. The IDE breakage is certainly annoying, but not a deal breaker if having the common source set makes our lives easier otherwise. But I'm worried that the Kotlin docs explicitly call this out as an unsupported combination. That makes me think there might be other gotchas we just haven't run into yet.

JeroenMols replied:
An alternative approach here could be to create two modules: `:runtime` and `:runtime-common`.

We would then extract all the `Base64` stuff out of `:runtime-common` and this module would only only have the JVM compile target (not Android). This would also allow to get rid of the `:jvm-test-types` module

`:runtime` would then offer the Android and `Base64` stuff on top.  and `:runtime` would then consume `runtime-common` as a dependency.

I don't think this is especially easy to do (because `Base64` is tangled with the rest), but might be worth exploring.

As for the IDE autocomplete: that's a good point, we could also avoid this by duplicating all code across Android/JVM but that would obviously have it's own drawbacks.

How would you like to tackle this moving forward? I can try out the two module approach? Or would you defer this to a follow up ticket?

garyp replied:
I suspect the two module approach is going to be complex to implement, as you say. However, I just remembered about a workaround that we were using in pbandk with Kotlin 1.3 for shared source sets (before HPP support was added) that might work here too:

```kt
    jvm {
        val main by compilations.getting {
            defaultSourceSet {
                kotlin.srcDir("src/commonJvmAndroid/kotlin")
            }
    }
    android {
        val main by compilations.getting { // this probably needs to be adjusted to use Android's sourceset configuration syntax
            defaultSourceSet {
                kotlin.srcDir("src/commonJvmAndroid/kotlin")
            }
    }
```

In essence, not creating a separate sourceset for `commonJvmAndroid` but instead having both JVM and Android sourcesets reuse the source files from that directory. Can you see if that could work?

JeroenMols replied:
This was a great suggestion! 🙏 

I tested in Android Studio and all imports are resolved + build works from command line. Could you please double check the same in Intellij?

garyp replied:
Seems to be working in IntelliJ too 👍

JeroenMols commented on runtime/build.gradle.kts:
```diff
@@ -50,7 +60,19 @@ kotlin {
+            dependsOn(commonMain)
+        }
+
+        val androidMain by getting {
```
Android sourceSet that contains a `Util` class and a `AndroidManifest`

JeroenMols commented on runtime/build.gradle.kts:
```diff
@@ -91,6 +113,15 @@ kotlin {
     }
 }
 
+android {
```
Note: we unfortunately have to be opinionated about a min and target Android version here...

JeroenMols commented on runtime/build.gradle.kts:
```diff
@@ -91,6 +113,15 @@ kotlin {
 
+android {
+    compileSdkVersion(Versions.androidTargetSdk)
+    sourceSets["main"].manifest.srcFile("src/androidMain/AndroidManifest.xml")
```
From Kotlin multiplatform example: https://github.com/Kotlin/kmm-sample/blob/325f92e845b4186d0ea82247aa7143018831aa1f/shared/build.gradle.kts#L51

JeroenMols commented on runtime/src/androidMain/AndroidManifest.xml:
```diff
@@ -0,0 +1,2 @@
+<?xml version="1.0" encoding="utf-8"?>
+<manifest xmlns:android="http://schemas.android.com/apk/res/android" package="pro.streem.pbandk"/>
```
`pbandk` wasn't allowed as package name, so took your full url

JeroenMols commented on runtime/src/androidMain/kotlin/pbandk/internal/Util.kt:
```diff
@@ -0,0 +1,8 @@
+
+import android.util.Base64
+
+internal actual object Util : AbstractUtil() {
```
This is the actual fix

JeroenMols commented:
Build failure is caused by the removal of `withJava()`: this option is needed to run the unit tests which are still in Java, but cannot be applied at the same time as the Android plugin as that would cause a build error:

```
The 'java' plugin has been applied, but it is not compatible with the Android plugins.
```

Basically it's impossible to have both the Android and Java plugin on a project at the same time:
- Jetbrains issue: https://youtrack.jetbrains.com/issue/KT-30878
- Android issue: https://issuetracker.google.com/issues/132576775

Possible solutions:
- Convert tests to `Java`
- Add build flag to toggle between plugins -> all `Android` blocks would need to be put behind an if statement (not that straight forward)
- Extract all Java code into a separate Java only module

JeroenMols commented:
Looked into Kotlin conversion, but this is close to impossible due to most java files being longer than 30k lines.

JeroenMols commented:
Moved all Java code for tests into a separate module `jvm-test-helpers`. This makes it possible to apply the Java plugin without conflicting with the Android plugin + has the added benefit of isolating all Java code into a separate module.

This solution has some overhead however: an extra module.

Further I also had to add a new sourceSet `androidTest` where I provide the JVM implementation of `pbandk.internal.Util`, because the Android variant doesn't work during unit tests and throws a `NotMockedException`.

All in all, this solution turned out to be a bit hairier that I had hoped, but I tried to make the best out of all complexity.

garyp reviewed (commented):
Some initial thoughts...

JeroenMols commented on build.gradle.kts:
```diff
@@ -4,6 +4,16 @@ import org.gradle.api.tasks.testing.logging.TestLogEvent
+    }
+
+    dependencies {
+        classpath("com.android.tools.build:gradle:${Versions.androidGradlePlugin}")
```
Thanks for the tip, I'll look into this!

JeroenMols replied:
Done, for some weird reason I couldn't use a `when` statement... 🤔  so I ended up using an if/else.

garyp replied:
Oh right, we're still using Groovy rather than Kotlin for `settings.gradle`. Forgot to convert that one over to Kotlin when we did the other gradle files.

JeroenMols commented:
Ready for a new round of reviews!

garyp commented on runtime/src/androidTest/kotlin/pbandk/internal/Util.kt:
```diff
@@ -0,0 +1,12 @@
+
+/**
+ * Same implementation as JVM to run unit tests on Android as [android.util.Base64] will throw
+ * a NotMockedException.
```
Hmm. The downside of this approach is that the unit tests don't test the actual Android implementation. For example, it's possible that Android's `Base64` might handle edge cases (such as support for Base64Url, line wrapping, etc.) differently from Java's `Base64`. But there's no way to verify that by running the unit tests 😞 

Unfortunately I don't have any great suggestions here. The only thing I can think of is to run the unit tests on Android using Roboelectric. But from some initial googling, I don't see how we can configure the Kotlin Multiplatform tests that come from `commonTest` to use the Roboelectric test runner for the `android` platform.

garyp replied:
I think I got something that works! Add Roboelectric's Android stubs as a runtime test dependency:

```kt
        val androidTest by getting {
            dependencies {
                ...
                runtimeOnly("org.robolectric:android-all:11-robolectric-6757853")
            }
        }
```

Then delete the `androidTest` version of `Util.kt`. The tests should now run correctly with the Android Base64 library. There's one test with a legitimate failure that I'll leave up to you to fix 😉

JeroenMols replied:
If I use Robolectric `v11-...` then I get the error:

```
java.lang.UnsupportedClassVersionError: android/util/Base64 has been compiled by a more recent version of the Java Runtime (class file version 53.0), this version of the Java Runtime only recognizes class file versions up to 52.0
```

And if I use Robolectric `v10-...` (october 2019) I get the error: 

```
java.lang.AssertionError: expected:<{"optionalBytes":"AQI="}> but was:<{"optionalBytes":"AQI=\n"}>
```

Two thoughts about the failure:

- The version of `Robolectric` used is already almost 1.5 years old (installing a newer version of the Java SDK might solve this though)
- Is this failure representative for Android or is this a bug in `Robolectric`

--------------------------------------

Few thoughts on using Robolectric:

1. The platform specific `Base64` implementation wan't tested before
before adding the Android artifact, we were using the Android `java.util.Base64` which was already a different implementation from the JDK `java.util.Base64`

2. Robolectric is only an approximation of the Android framework
imagine that the tests pass using Robolectric, how much guarantee does that give for an actual Android device? In the end Robolectric has some version of `android.util.Base64` hardcoded into it, potentially even modified so it works run during unit tests.

3. What value would such a test add?
even if the `android.util.Base64` implementation of Robolectric is exactly the same as the one in the Android platform, then this test only serves to prove that `android.util.Base64` works. 
Shouldn't that be the responsibility of the Android team? Note also that this class exists since API level 1, has a proven track record and has only been [modified 6 times the past 10 years](https://cs.android.com/android/platform/superproject/+/master:frameworks/base/core/java/android/util/Base64.java;l=29;bpv=1). 
We aren't going to test Android's String implementation either, right? (in this regard I highly recommend this article: https://www.facebook.com/notes/kent-beck/unit-tests/1726369154062608/)

------------------------
So while everything can be tested, not all tests are created equal!
Sometimes the value added isn't proportional to the effort it took writing and maintaining such a test.

To me personally Robolectric is one such case: I've honestly completely given up on that framework. It slows my tests down, is a pain to maintain and most important of all it doesn't spark joy.

-----------------------
That being said, if we choose to use Robolectric, I would highly appreciate your advice on to fix:

```
java.lang.AssertionError: expected:<{"optionalBytes":"AQI="}> but was:<{"optionalBytes":"AQI=\n"}>
```

JeroenMols replied:
The actual implementation would be a bit hairier than this as `Build.VERSION.SDK_INT` is 0 during unit tests:

```kotlin
actual override fun base64ToBytes(str: String): ByteArray =
    if (doesntSupportJavaBase64()) {
        android.util.Base64.decode(str, android.util.Base64.DEFAULT)
    } else {
        java.util.Base64.getDecoder().decode(str)
    }
    
// Also need to check for minimum version as [SDK_INT] is 0 during unit tests
    private inline fun doesntSupportJavaBase64() =
        Build.VERSION.SDK_INT > Build.VERSION_CODES.LOLLIPOP && Build.VERSION.SDK_INT < Build.VERSION_CODES.O
  ```
   
Begs the question that try/catching wouldn't be better in this case:

```kotlin
actual override fun base64ToBytes(str: String): ByteArray = try {
    java.util.Base64.getDecoder().decode(str)
} catch (e: Throwable) {
    // Fallback to Android implementation on platforms below API 26
    android.util.Base64.decode(str, android.util.Base64.DEFAULT)
}
```

JeroenMols replied:
I've added such an implementation, not to push this either way, but just to show you how that would look like.

Please advice on how you prefer to proceed.

garyp replied:
I should've been more specific in my comment yesterday. This:

```
java.lang.AssertionError: expected:<{"optionalBytes":"AQI="}> but was:<{"optionalBytes":"AQI=\n"}>
```

is an actual bug in pbandk that was found by running the tests with Robolectric. It's a bug that would've occurred on a real Android device. The purpose of running unit tests with Robolectric is not to guard against bugs in Android's implementation of standard Java APIs. I agree with you that that would be overkill and completely unnecessary. We can assume Android's implementation is well-tested. The purpose of running unit tests with Robolectric is to guard against bugs where pbandk's `android` variant uses different APIs from the `jvm` variant (such as with Base64), which might work subtly differently. Adding an API check to the code, as you're suggesting, doesn't fix the problem. You'll still have a bug on devices with an API level below `Build.VERSION_CODES.O`.

It looks like the cause of this bug is because `android.util.Base64` by default line wraps the Base64 output. Whereas `java.util.Base64` does not. No line wrapping is the behavior that other protobuf libraries will expect. I think changing the code to this will fix the bug:

```kt
Base64.encodeToString(bytes, Base64.NO_WRAP)
```

garyp replied:
Regarding the error you get when using the `v11` Robolectric library:

```
java.lang.UnsupportedClassVersionError: android/util/Base64 has been compiled by a more recent version of the Java Runtime (class file version 53.0), this version of the Java Runtime only recognizes class file versions up to 52.0
```

I suspect this is because you're running the tests with Java 8 on your local machine (I think this is what Android Studio bundles). Whereas it looks like the Robolectric `v11` library was compiled with Java 9. I'm running the tests under Java 15 on my machine, which is why I didn't hit this issue. I'm fine with using `v10` for now to maintain Java 8 compatibility. I think Android is planning to update to Java 11 in the next Android Studio release.

garyp commented:
Related to the Android unit testing question: how can we run the conformance tests for the android variant? I wonder if Roboelectric could be leveraged somehow there too. Alternatively, I just came across https://github.com/ReactiveCircus/android-emulator-runner, which makes it easy to run the Android Emulator in GitHub Actions. So another route could be to spin up an emulator, push some files to it, and use `adb shell` to run the conformance tests on the emulator.

This doesn't have to block this PR by the way. I'm just thinking out loud. We can create new GitHub issues to track this tech debt.

garyp commented:
Oh, one more thing: can you please update `README.md` with Android-related info where needed, and add an entry to `CHANGELOG.md`.

JeroenMols commented on buildSrc/src/main/kotlin/Versions.kt:
```diff
@@ -8,4 +8,7 @@ object Versions {
     const val springBootGradlePlugin = "2.3.7.RELEASE"
     const val osDetectorGradlePlugin = "1.6.2"
+    const val androidGradlePlugin = "4.1.2"
+    const val androidMinSdk = 14
```
We require min SDK 21, so I'm fine with that.

FYI: did a quick search to see what others do:
- Stripe is min SDK 24
- Square is min SDK 21
- Coil is min SDK 14
- Glide is min SDK 14
- Lottie is min SDK 16

JeroenMols replied:
```suggestion
    const val androidMinSdk = 21
```

JeroenMols commented:
Processed feedback, updated `README.md` and `CHANGELOG.md` and filed ticket to tackle conformance tests https://github.com/streem/pbandk/issues/128.

Ready for another review

JeroenMols commented:
Update the tests and fixed the Android Base64

garyp commented on runtime/build.gradle.kts:
```diff
@@ -50,14 +60,32 @@ kotlin {
+                implementation(kotlin("test"))
+                implementation(kotlin("test-junit"))
+                implementation("junit:junit:4.12")
+                runtimeOnly("org.robolectric:android-all:10-robolectric-5803371")
```
Please add a comment that we shouldn't bump this version higher than `10` until Android Studio is using Java 11 instead of Java 8.

JeroenMols replied:
I'll also move this version to `Versions.kt`

garyp commented on CHANGELOG.md:
```diff
@@ -17,6 +17,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 * `MessageDescriptor` (available via `Message.descriptor` or `Message.Companion.descriptor`) is now part of the public API. Initially only `MessageDescriptor.fields` is public, which provides access to descriptors for all of the message's fields. Additional properties of the message descriptor will be exposed in future versions. Please file an issue on GitHub if there are specific properties you  ...
 * `FieldDescriptor` (available via `MessageDescriptor.fields`) is now part of the public API. Initially, the field's `name` and `options` are public, allowing access to custom options defined on the field in the `.proto` file. The field's `value` accessor is also public, though this API will probably change before the final 0.10.0 release. (PR [#103])
 * Add the ability to unit test the output of `CodeGenerator` (PR [#117], fixes [#48]) (thanks @nbaztec)
+* Added a dedicated Android artifact `pbandk-runtime-android` of the `PBandK` runtime. To migrate to this new artifact, replace `implementation("pro.streem.pbandk:pbandk-runtime-jvm:0.9.1")` with `implementation("pro.streem.pbandk:pbandk-runtime-jvm:0.10.0")` (PR [#124], fixes [#75]) (thanks @jeroenmols)
```
I think you meant to use `pbandk-runtime-android` in the replacement rather than `pbandk-runtime-jvm` 😉

JeroenMols replied:
Good catch!

garyp commented on runtime/src/androidMain/kotlin/pbandk/internal/Util.kt:
```diff
@@ -0,0 +1,11 @@
+        Base64.decode(str, Base64.DEFAULT)
+
+    actual override fun bytesToBase64(bytes: ByteArray): String =
+        Base64.encodeToString(bytes, Base64.DEFAULT)
```
I think you forgot to update this based on the failing test

JeroenMols replied:
Wow my, messed up the commits after wanting to split them up 🤦

garyp reviewed (changes requested):
A couple final things and then should be good to merge

JeroenMols commented:
Fixed failing tests, updated changelog, added comment about updating Robolectric and moved versions to Versions.kt.

Ready for a new round of reviews

garyp reviewed (approved):
Thanks again for working on this! 👍
